content/en/blog/extend-falco-inputs-with-a-plugin-the-basics.md

```diff
@@ -0,0 +1,670 @@
+	"github.com/falcosecurity/plugin-sdk-go/pkg/sdk/plugins/extractor"
+	"github.com/falcosecurity/plugin-sdk-go/pkg/sdk/plugins/source"
+
+	dockerTypes "github.com/docker/docker/api/types"
+	dockerEvents "github.com/docker/docker/api/types/events"
+	docker "github.com/moby/docker/client"
+)
+```
+
+In almost everybody plugins we'll write, these different component of `plugin-sdk-go` we'll be imported, they're really convenient and provide a much easier way to deal with `Falco plugin framework`.
+
+### The structures
+
+2 structures are mandatory and must respect `interfaces` of the `sdk`:
+
+```go
+// DockerPlugin represents our plugin
+type DockerPlugin struct {
+	plugins.BasePlugin
+	FlushInterval uint64 `json:"flush_interval" jsonschema:"description=Flush Interval in seconds (Default: 2)"`
+}
+
+// DockerInstance represents a opened stream based on our Plugin
+type DockerInstance struct {
+	source.BaseInstance
+}
+```
+
+* `DockerPlugin` represents our plugin that will be loaded by the framework. Embedding `plugins.BasePlugin` allows to respect the [`Plugin interface`](https://pkg.go.dev/github.com/falcosecurity/plugin-sdk-go@v0.1.0/pkg/sdk/plugins#Plugin) of the `sdk`
+* `DockerInstance` represents a stream of events opened by the framework with the plugin. Embedding `source.BaseInstance` allows to respect the [`Instance interface`](https://pkg.go.dev/github.com/falcosecurity/plugin-sdk-go@v0.1.0/pkg/sdk/plugins/source#Instance) of the `sdk`.
+
+
+For both `structs` we can extra attributes for any purpose we need, like for configuration. In this example we have `FlushInterval` which represents the frequency events will be sent to the `instance` by the `docker` client we'll create. This attribute will have a default value that can be overridden by `init_config` in `plugins` section, see []().
+
+### The functions and methods
+
+#### `main()`
+
+The `main()` funcion is mandatoy as for any `go` program but because we'll build the `plugin` as a library for the `Falco plugin framework` which is written in `C`, we can let it empty.
+
+```go
+// main is mandatory but empty, because the plugin will be used as C library by Falco plugin framework
+func main() {}
+```
+
+#### `init()`
+
+The `init()` function is used for regestering our plugin to the `Falco plugin framework`, as a [`source`](https://falco.org/docs/plugins/#source-plugin) and and [`extractor`](https://falco.org/docs/plugins/#extractor-plugin).
+
+```go
+// init function is used for referencing our plugin to the Falco plugin framework
+func init() {
+	p := &DockerPlugin{}
+	extractor.Register(p)
+	source.Register(p)
+}
+```
+
+#### `Info()`
+
+This method is mandatory, all plugins must respect, it allows the `Falco plugin framework` to have all intels about the plugin itself:
+
+```go
+// Info displays information of the plugin to Falco plugin framework
+func (dockerPlugin *DockerPlugin) Info() *plugins.Info {
+	return &plugins.Info{
+		ID:                 6,
```

Comment:
Will you register this plugin?

Reply:
not 100% I will, I need to test it with docker swarm first.